Byzantine Fault Tolerant block replication and block signature verification (#48)
hyperledger/fabric-rfcs: text/0010-bft-signatures.md

and there is currently no design of how blocks are to be retrieved in the Byzantine setting or how signatures on blocks should be verified.

Block retrieval from Byzantine orderers needs to tackle denial of service from malicious orderers.

The challenge with block signature validation in the Byzantine setting is avoiding the linear space overhead in the number of certificates that need to be stored as part of the signature information.

While forks of Fabric with a Byzantine Fault Tolerant ordering service [exist](https://github.com/SmartBFT-Go/fabric/), they have not been incorporated into the official Hyperledger Fabric project.

This RFC aims to:

- Give an overview of how denial of service resistant block retrieval and block signature validation is done in SmartBFT.
- Define how denial of service resistant block retrieval should be conducted in Hyperledger Fabric.
- Define how Byzantine block signature validation should be constructed and encoded in Hyperledger Fabric.

# Motivation
[motivation]: #motivation

Currently, Hyperledger Fabric only supports ordering service types that are Crash Fault Tolerant (CFT), not Byzantine Fault Tolerant (BFT).

> There are two main attacks that are considered outside of the crash fault tolerant thread model, but must be protected against in the Byzantine fault tolerant setting:

Reviewer: Pedantic point, but is the word 'thread' here significant? i.e. is there a difference between multi and  single-threaded model of execution that has relavent impact?

Author: Typo. It's supposed to be threat, of course

**Block withholding**: A Byzantine orderer might be reluctant to send newly formed blocks to peers that connect to it, either because of malicious intent or to save egress network bandwidth charges from its hosting provider.

>  Fabric architecture relies on peers receiving newly formed blocks as fast as possible, because otherwise client transaction submission times out and smart contract simulation results in stale reads.

Reviewer: or applying Hanlan's Razor, the nodes are overloaded for example.

The time frame is not defined here, - is this the 'suspicion threshold' mentioned later?

Author: No

**Block forgery**: A Byzantine orderer might try to forge a block and send it to peers or other ordering nodes that replicate from it, in order to alter the smart contract execution on peers by subterfuge.
 One way of doing this, is to omit a transaction from a block, or re-order transactions within a block.
 
To circumvent block forgery attacks, the following invariant should be preserved: *A set of signatures on a block should only be considered valid if it is impossible for the parties running the protocol to construct a set of signatures on a different block that is also considered valid*.

The aforementioned invariant can be implemented in different ways, such as requiring a **threshold of signatures** from different orderers on a block, or requiring a single **threshold signature** on a block.

> The threshold itself can vary depending on the consensus algorithm and on the stage in it that the signatures is produced: Signing a block post consensus requires signatures from only `f+1` orderers (at the expense of latency), while signing during consensus requires signatures from `2f+1` orderers (at the expense of size and computation).

Reviewer: For my own information and learning - could you point me to a reference that explains the difference here please? Why `f+1` and `2f+1`?

Author: It follows from the BFT protocol: If the BFT protocol has three stages, like PBFT, then the last stage involves collecting `2f+1` messages, each containing a hash on the block, and all `2f+1` messages need to be with the same hash. 

Since you want to transfer this block to someone else, that someone else needs to be sure that you really collected `2f+1` messages, it cannot take your word for it. So you make each message signed. Now when you send this block to someone else, it can verify the signatures and see that you really received the messages in the third phase from `2f+1` different nodes, so that block was produced in a sound execution of BFT.

But there is another way of collecting signatures that requires only `f+1` signatures. Consider a BFT protocol that doesn't use signatures at all, and after each block is committed, the nodes simply send each other signatures over the blocks they have agreed upon during the BFT protocol. If you have `f+1` signatures over such blocks, then it means that you have at least a single honest node (since up to `f` nodes are malicious) that participated in the BFT protocol and committed that block. From the correctness of the BFT protocol, it follows that it is the only possible block that can be agreed upon, and since you now know that an honest node committed that block, then you can also trust that block yourself.

# Naive approaches for block replication and block signature verification

### Block replication

In a Byzantine setting, there is a fundamental assumption that there is an upper bound on the number of malicious or failed nodes
 among all the nodes that run the protocol, and that upper bound is denoted as `f`.

>  In BFT, the ordering service is comprised of `n>3f` for some `f>0`, so we have up to `f` nodes which are either offline, unreachable or malicious,

Reviewer: Is it correct that `n` has to be odd?

Author: `n` is odd when `f` is even and vice versa.

Reviewer 2: > `n` is odd when `f` is even and vice versa.

if one sets _f_ possibly high, which perhaps makes the most sense. However, in principle, one can, say, set _f=1_ and _n=5_.

Reviewer 2: > _f=1_ and _n=5_

In fact, this seems the only possible configuration if there are 5 independent stakeholders, each wanting to operate a PBFT node.

Author: Indeed, I am often simplifying by assuming `n=3f+1`. 

In reality the number of messages you collect instead of `2f+1` is from a "quorum" of nodes, and it's defined in a way that every two quorums have at least a single correct node.

>  In BFT, the ordering service is comprised of `n=3f+1` for some `f>0`, so we have up to `f` nodes which are either offline, unreachable or malicious,

Reviewer 2: ```suggestion
 In BFT, the ordering service is comprised of `n>3f` for some `f>0`, so we have up to `f` nodes which are either offline, unreachable or malicious,
```

Author: While it's true that it is comprised of `n>3f`, in SmartBFT you don't get to pick `f` but only `n`, and from `n` - the system automatically picks `f`. 

I believe this is the right approach as you obviously cannot quantify `f` as you don't know upfront which servers can be faulty at the same time in the first place.

Reviewer 2: > you don't get to pick `f` but only `n`

Then it must be either `3f+1` or `3f+2` or `3f+3`, depending on what `n mod 3` is.

Author: Yeah indeed if `n` is not congruent to 1 modulo 3 then `f` is "smaller". In any case I accepted your edit suggestions, thanks for the corrections.

>  and `2f+1` nodes that are online, reachable and follow the protocol.

Reviewer 2: ```suggestion
 and at least `n-f` nodes that are online, reachable and follow the protocol.
```

```
message MetadataSignature {
    bytes signature_header = 1; // An encoded SignatureHeader
    bytes signature = 2;       // The signature over the concatenation of the Metadata value bytes, signatureHeader, and block header
}
```

The signature header contains the identity of the signer which carries an x509 certificate and the MSP identity that one of its certificate authorities issued that certificate:

```
message SignatureHeader {
    // Creator of the message, a marshaled msp.SerializedIdentity
    bytes creator = 1;

    // Arbitrary number that may only be used once. Can be used to detect replay attacks.
    bytes nonce = 2;
}
```

This means that when signed by `2f+1` (or `f+1`) orderers, each block contains a number of signatures that is linear in the number of ordering service nodes that participate in the consensus protocol.
When considering an ordering service that can tolerat two failures (`f=2`, hence `n=7`) this means that each block carries  3.5KB of x509 certificates,
even if the block itself contains only a single transaction (that is often 3.5KB in size itself).

> Clearly, encoding identities each time in the signature header of each orderer signature is wasteful and therefore we should strive to eliminate this overhead.

Reviewer 3: It is a tradeoff between storage and network bandwidth+verification speed.
Another problem with the static signature collection is that the orderers set in the channel can change, which will make the signature collection difficult to be applicable always.

## Existing approaches for block replication and block signature verification
[Existing approaches]: #existing-approaches-for-block-replication-and-block-signature-verification

### Attestation for block existence

> An approach for block denial of service resistance taken by [SmartBFT](https://arxiv.org/abs/2107.06922) is to connect to (at least)`f+1` ordering service nodes, fetch blocks only from one of them and fetch attestations that blocks exist from the remaining `f` (or more).

Reviewer: Just to clarify, is it correct to that the block attestations fetched have to be for exactly the same blocks fully obtained?

Author: Well if you have a signature on the block header then there is no way of passing the verification check unless somehow the consensus protocol itself has a bug and leads to a fork.

> If a block existence attestation for a sequence **i** is found valid but the orderer designated to send blocks withholds block **i** for too long, the role assignment of block source and attestation for block existence source is re-shuffled.

Reviewer: What's the definition of  `sequence i`?  Is it a description of the blocks that are required?

Too long == suspicion threshold?

Author: A block number. 

And yes, suspicion threshold.

```

Let's consider the two possible outcomes based on the orderer that is selected for block retrieval:

1. If orderer that blocks are pulled from is withholding newly formed blocks, then there is at least one honest orderer that will not withhold attestations of block existence. Hence, a re-shuffling of role assignment will occur until an honest orderer will be the block source.
2. Else, the orderer that blocks are pulled from is not withholding newly formed blocks. However, it is possible that new blocks are not formed due to lack of client activity. It is imperative to ensure that malicious ordering nodes cannot make honest ordering nodes be falsely suspected of block withholding.
As each "block existence attestation" contains `2f+1` signatures it is impossible to forge one. Hence, if a node sends a attestation for block existence, it means this a block with such sequence number was indeed appended to the blockchain.

In order for a peer or an orderer to convey the intent of receiving only blocks without transactions, 
the [SeekInfo](https://github.com/hyperledger/fabric-protos/blob/main/orderer/ab.proto#L45) message is expanded to contain also a `SeekContentType` field:

```

>   // SeekContentType indicates what type of content to deliver in response to a request. If BLOCK is specified,

Reviewer 4: Given the current pattern of deliver service interfaces in [order](https://github.com/hyperledger/fabric-protos/blob/main/orderer/ab.proto) and [peer](https://github.com/hyperledger/fabric-protos/blob/6e299ac721985b2f616065aca4d0cb63dffeca40/peer/events.proto#L65), it makes better sense to define a separate service instead of fitting in `SeekInfo`

To take a note for the implementation level, we persist `header` and `metadata` separate from `transactions` in the block storage. So, we can add an API in blockstorage to just retrieve these two to save I/O.

Author: You can add an API in the block storage regardless, and just call it in the deliver server. 
I don't see why it's a problem that a peer will also be able to service this API. 
After all, it's the same common deliver server code in both node types.

Reviewer 4: May be you did not get the main suggestion here... let me paste again 

Given the current pattern of deliver service interfaces in [order](https://github.com/hyperledger/fabric-protos/blob/main/orderer/ab.proto) and [peer](https://github.com/hyperledger/fabric-protos/blob/6e299ac721985b2f616065aca4d0cb63dffeca40/peer/events.proto#L65), it makes better sense to define a separate service instead of fitting in `SeekInfo`

Author: I defined a separate gRPC service as you suggested.

>   // the orderer will stream blocks back to the peer. This is the default behavior. If HEADER is  specified, the

Reviewer 5: ```suggestion
  // the orderer will stream blocks back to the peer. This is the default behavior. If HEADER_WITH_SIG is  specified, the
```

  // orderer will stream only a the header and the signature, and the payload field will be set to nil. This allows
  // the requester to ascertain that the respective signed block exists in the orderer (or cluster of orderers).
  enum SeekContentType {
      BLOCK = 0;

>       HEADER_WITH_SIG =1;

Reviewer 4: BTW, I am not sure if I understand the significance of the header and signature here as the actual block would have `enough` signatures in it and as I understood, this deliver service is just to get to know that a new block has been committed. If that's the case, only block number should suffice?

Author: If you only have block numbers, you need to collect `f+1` attestations before understanding you have enough attestations for a block sequence. Any lower number and you might get fake attestations that will cause you to blame the current block source.

With signatures, you only need to collect one attestation, so the logic is simpler (no need to "count" and track sequences).

Reviewer 4: Yes, you will need to hear from `f+1` for block number (that anyways you would have to connect to either way). Not sure about the logic being simpler v/s difficult as even in the signature way, you would have to assume that f attestations may not be correct. I guess, it would be choice between perform signature matching over attestation (mostly one at run-time but code for handling from all failure) v/s just taking hint from `f+1`

Hence, while a malicious orderer may not withhold blocks indefinitely, it may slow down the replication of its consumers.

Another minor inconvenience is that the peer admin needs to configure the peer correctly to use the Byzantine fault tolerant implementation instead of the regular one.

### Signature identity referencing

In SmartBFT, like Raft, each configuration block contains a set of consenter (ordering service consensus protocol participant) definitions. 
Each such consenter definition contains various information that is used either to connect or to authenticate the consenter:

```
message Consenter {
    uint64 consenter_id = 1;
    string host = 2;
    uint32 port = 3;
    string msp_id = 4;
    bytes identity = 5;
    bytes client_tls_cert = 6;
    bytes server_tls_cert = 7;
}
```

When validating a signature signed by orderer with numerical identifier **j**, the mapping between **j** and the matching identity (which carries the public key) is taken from the latest configuration block.

> Recall, the naive approach to encode several signatures on a block in its metadata involves encoding an amount of orderer identities that is linear in the number of ordering nodes.

Reviewer 6: if I am correct, we avoid linear space in block bandwidth by `Consenter` structure, and we can't avoid linear space stored in peer/orderer, as we use x509 cert to verify msg(in BFT), as peer/orderer need translate from `Consenter` to find a specific x509 cert in memory/storage for verification works.

Author: Not sure if I understand what you're trying to say, but yes we fetch the certificate by the id (number) and then replace it in-memory and proceed as usual.

To that end, SmartBFT substitutes the identity (over 700B in size) with its consensus node identifier (a 64 bit number). 
As a result, the `MetadataSignature` now contains three additional fields:

```
message MetadataSignature {
    bytes signature_header = 1; // An encoded SignatureHeader
    bytes signature = 2;       // The signature over the concatenation of the Metadata value bytes, signatureHeader, and block header

>     uint64 signer_id = 3;

Reviewer 6: I am not sure if it is better from
```
    uint64 signer_id = 3;
```
to
```
    string signer_id(hash from x509 OU,ROLES, expire date etc... been used in endorsement policy and x509 cert management.)
```

Author: When you validate orderer signatures, expiration doesn't take effect. And signer ID is shorter than SHA256 hash, it has 8 bytes instead of 32 bytes.

    bytes nonce = 4;
}
```

When a Fabric node validates a signature on a block, an ephemeral `SignatureHeader` is constructed where the `nonce` is taken from the
`MetadataSignature` and the `creator` is the `identity` matching the `signer_id`. 

> 

Reviewer 4: Not related to this RFC along, in general, I know that we include the identity as part of the value for the signatures but not sure if enhances security by any means? Can we not simply live without it and hence the consenter mapping?

Author: The fact that we sign over the public key is irrelevant.

You still need to know which public key to use to validate the signature, right? When sending signatures and storing them on storage you need either to mark who are the signers or to simply attach the identity to the signature.

If you attach the identity to the signature, then you get a high storage overhead. 

If you don't attach the identity, then you need to know what identities are possible candidates. But where are you going to take them from? It's tempting to take the block signature policy and convert it to identities and then try to match a candidate MSP principal to a signature but what if the policy is going to be with role MSP principals? then you won't be able to do that anymore.

Reviewer 4: > The fact that we sign over the public key is irrelevant.

The relevance of this was, quite a good part part of you write up was around how to avoid including the cert and my question was can we get rid of this all together?

yes, of course you need the hint of Ids!

> This technique reduces the aforementioned overhead of 3.5KB to a meager 40 bytes.

Reviewer 6: Cool, and I want to see more details about `Consenter` lifecycle as the x509 cert been management. ex expire
and how to distribute a new signer across fabric network.

and BTW, how about Idmix?

Author: But details about consenter lifecycle are for another RFC, not for this one. This is an RFC about block replication and signature verification and encoding. 

Why would you have orderers sign with idemix? Do you want to hide which orderer signed? why?

Reviewer 6: May I know the details for "another RFC" ?
ok... we don't need have orderers sign with idemix, as we don't need to hide which orderer signed.

> This technique reduces the aforementioned overhead of 3.5KB to a meager 40 bytes.

Reviewer 5: 40 bytes per orderer, or 40 bytes total? Could you spell it out... what exactly are the 40 bytes?

Author: So if you have a cluster of typical size with `f=2` (supports two failures) your signature set size is 5 signatures. 
In the vanilla approach, that amounts to more than `700`x`5` = `3500` bytes of overhead because you need to put the entire x509 identity alongside each signature.

With identity referencing, you only need to have `8` bytes (uint64) for each signature, so it's a total of `40` bytes.

While the approach of the SmartBFT library integration works, it creates a coupling between the consensus implementation and the rest of Fabric, and this limits incorporation of other consensus implementations into Fabric.
 
More specifically, it requires peers to understand how to decode the consensus metadata of ordering service nodes. That means that the peer needs to be aware of the type of ordering service it is working with, which breaks the abstraction of the ordering service as a service that receives as input transactions, and outputs blocks.

# Proposed techniques
[technique]: #proposed-technique

The proposed approach is heavily inspired by SmartBFT but is designed to be consensus agnostic, minimize implementation and operational complexity.

### Block replication

> The proposed technique for block replication is essentially the approach of SmartBFT where a re-shuffling takes place not only upon suspicion of the block source but also in a timely fashion in order to be resilient to overloaded or deliberately slow orderers.   

Reviewer 5: What is the criteria for "suspicion of the block source"?

Author: If you suspect the orderer is deliberately not sending blocks to you, because you get headers from other orderers.

> In order to save network bandwidth and processing power when the network is idle, the automatic time-based re-shuffling will only take place if new blocks are received.

Reviewer 5: "the automatic time-based re-shuffling will only take place if new blocks are received"... I don't get it, I thought the intent of time-based re-shuffling was to distinguish between "no client activity" and "orderer is withholding blocks". So I would expect time-based re-shuffling would be done when peer doesn't receive any blocks in a timely manner, not when "new blocks are received".

Author: No. There are two things:

1. Suspicion of a block source, upon reception of headers from other orderers but no reception of blocks from the orderer that supplies blocks.
2. Periodical reshuffling, upon block reception. Why upon block reception? Because suppose you receive blocks at a rate of 2 blocks per second, but the orderer that sends you blocks, sends them at a rate of 1 block per two seconds. The mechanism of (1) will not kick in and you will not suspect that orderer. However you will still linger behind in block reception because of the difference between the rate at which blocks are produced and the rate at which you receive them. Periodically rotating the orderer that will send you blocks, will solve this because once you rotate to an honest orderer you will quickly receive the blocks that were created while you got the blocks too slowly from the first orderer.

Moreover, there will not be a **need** to specify in the peer's local configuration (or an orderer's local configuration) which of the implementations to use. 
Notice that a block validation policy that mandates signatures from multiple ordering nodes, is **never** satisfied by **only** principal sets of size one. 
Or, equivalently, a block validation policy that is satisfied by **only** principal sets of size one never requires signatures from multiple ordering nodes. 
Evidently, policies of the former type are **always** BFT policies, while policies of the latter type might be either a crash fault policy or a BFT policy that utilizes a threshold signature scheme.

> Therefore, Fabric nodes will pick the BFT implementation if either the block validation policy implies the need, or if an override is activated in the node's local configuration.  

Reviewer 4: The above para gives an impression that you are proposing this technique of connecting with `f+1` as a general mechanism so as to avoid the need of a local configuration. But in the last statement, you mention `if an override is activated in the node's local configuration`. Are you proposing a local configuration here? Also, it's worth mentioning if the peer is expected to derive the value of `f` (for enable connecting to `f+1`) from the total numbers of OSNs mentioned in the channel config.

Author: It is indeed expected to derive it from there, according to the threshold of the policy, unless overriden in the local config.

### Block signature verification

The proposed technique for block signature verification should be generic and consensus agnostic, in order to facilitate incorporation of multiple consensus types without the need for Fabric nodes to understand which consensus implementation is employed.

Apart from the already existing block validation policy:

```
# BlockValidation specifies what signatures must be included in the block
# from the orderer for the peer to validate it.
BlockValidation:
    Type: ImplicitMeta
    Rule: "ANY Writers"
```

> 

Reviewer 5: Can you provide an example of BlockValidation policy in the case of SmartBFT?

Author: So in the original SmartBFT, the block validation policy was actually computed dynamically ([implicitly](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/common/policies.proto#L16-L27)).  In short - the node would look at the last configuration block, and compose a policy that is made of [identity](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/msp/msp_principal.proto#L45) principals according to the [identities](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/orderer/smartbft/configuration.proto#L25) of the consenters and would require a [threshold](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/common/policies.proto#L42-L43) of `2f+1` signatures for the policy to be satisfied. 

In the proposed design, the policy can simply be a regular signature policy (not an implicit one!) that will have identities in it. However, when signatures will be encoded, they will be encoded using the references of the consenters. Then, when the signature is to be evaluated, the reference will be used to look up the identity (from the consenter mapping) and then the regular policy engine will be activated, over the regular signature policy in a manner that is transparent to the signature policy engine.

The `Orderer` section in the channel config will contain a mapping from numerical identifiers to tuples of serialized identities:

```
Orderer: &OrdererDefaults
    ...
    ConsenterMapping:
    - Id: 1
      MSPID: OrdererOrg1
      Identity: <base64 encoding of raw identity>
    - Id: 2
      MSPID: OrdererOrg2
      Identity: <base64 encoding of raw identity>
      ...
```

Upon a configuration update, it will be up to the (Fabric side of the) consensus implementation to validate that:

- The block validation policy matches the required threshold (e.g `2f+1`) of the consensus protocol.

> - The `ConsenterMapping` is not in conflict with the opaque consensus metadata.

Reviewer 5: What exactly is the "opaque consensus metadata"?

Author: It's an encoding of the consensus-specific configuration in the orderer section of the config block. It contains [these items](https://github.com/hyperledger/fabric-protos/blob/main/orderer/etcdraft/configuration.proto#L14-L17) in Raft and similarly [these items](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/orderer/smartbft/configuration.proto#L14-L17) in SmartBFT.

The problem of correctly composing the consenter mapping and block validation policy at bootstrapping is considered out of scope of this RFC,

> as this is consensus specific and can be addressed with custom tooling, such as `configtxgen` enhancements or any other tooling.

Reviewer 5: I understand the details are out of scope, but could you at least provide a simple example of block validation policy and consenter mapping for SmartBFT so that we understand the intent?

Author: Yeah, for a 4 sized cluster of orderers 1 to 4, a block validation policy that's a [signature policy](https://github.com/hyperledger/fabric-protos/blob/main/common/policies.proto#L28-L32), and its [rule](https://github.com/hyperledger/fabric-protos/blob/main/common/policies.proto#L41-L43) is 3 out of 4 signatures out of the following MSP principals:

```
msp.MSPPrincipal{
				PrincipalClassification: msp.MSPPrincipal_IDENTITY,
				Principal:               protoutil.MarshalOrPanic(&msp.SerializedIdentity{IdBytes: []byte(orderer1 PEM content), Mspid: Org1MSP}),
			}, 
msp.MSPPrincipal{
				PrincipalClassification: msp.MSPPrincipal_IDENTITY,
				Principal:               protoutil.MarshalOrPanic(&msp.SerializedIdentity{IdBytes: []byte(orderer2 PEM content), Mspid: Org2MSP}),
			},
msp.MSPPrincipal{
				PrincipalClassification: msp.MSPPrincipal_IDENTITY,
				Principal:               protoutil.MarshalOrPanic(&msp.SerializedIdentity{IdBytes: []byte(orderer3 PEM content), Mspid: Org3MSP}),
			},
msp.MSPPrincipal{
				PrincipalClassification: msp.MSPPrincipal_IDENTITY,
				Principal:               protoutil.MarshalOrPanic(&msp.SerializedIdentity{IdBytes: []byte(orderer4 PEM content), Mspid: Org4MSP}),
			}
```

> The `MetadataSignature` shall be expanded to contain a `uint32` field of `signer_id` and the `bytes` field of the nonce` as in SmartBFT.

Reviewer 6: missing ` here?

nonce` 

`nonce`

- **Common deliver service**: The [common deliver service](https://github.com/hyperledger/fabric/blob/main/common/deliver/deliver.go), used by both peers and orderers when sending blocks to subscribers, will be enhanced by an ability to 
redact the transactions from a block prior to its dispatch based on the intent derived from the `SeekInfo` message as done in SmartBFT.

This RFC suggests to build a re-usable component for block pulling that will be used by both the peer and the orderer, and it will have a simple API as the [BlockPuller](https://github.com/hyperledger/fabric/blob/main/orderer/common/cluster/deliver.go#L30-L55):

```
// BlockFetcher is used to pull blocks from ordering service nodes
type BlockFetcher interface {
	PullBlock(seq uint64) *common.Block
	Close()
}
```

The `BlockFetcher` component will perform all bookkeeping required for both CFT and BFT use cases and will then be incorporated in the orderer and peer block replication code.
This will let the project get rid of multiple implementations that perform the same thing, as currently there is a duplicity in the form of the peer deliver client and the orderer's cluster infrastructure.

> 

Reviewer 6: Not sure if we need to considering changing SDK/peer endorsement service?

for ex
when any signer submit a Tx.
if the signer x509 information been "cached/stored" in Peer.
return `signer_id(singer identifier information)` in `SignedProposal`
else
keep return x509 information?

as a part of singer identifier information management?

Author: This is out of the scope of the RFC.
